File: early/46.permutations.py
#
# [46] Permutations
#
# https://leetcode.com/problems/permutations
#
# Medium (42.74%)
# Total Accepted:    164487
# Total Submissions: 384821
# Testcase Example:  '[1,2,3]'
#
# 
# Given a collection of distinct numbers, return all possible permutations.
# 
# 
# 
# For example,
# [1,2,3] have the following permutations:
# 
# [
# ⁠ [1,2,3],
# ⁠ [1,3,2],
# ⁠ [2,1,3],
# ⁠ [2,3,1],
# ⁠ [3,1,2],
# ⁠ [3,2,1]
# ]
# 
# 
#
class Solution(object):
    def permuteRecurse(self, nums, index, all):
        if index == len(nums):
            all.append(nums[:])
            return 
        for i in range(index, len(nums)):
            a, b = nums[index], nums[i]
            nums[index], nums[i] = nums[i], nums[index]
            self.permuteRecurse(nums, index + 1, all)
            nums[index], nums[i] = a, b
        
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        all = []
        self.permuteRecurse(nums, 0, all)
        
        return all

File: 2020/46.permutations.cpp
/*
 * @lc app=leetcode id=46 lang=cpp
 *
 * [46] Permutations
 *
 * https://leetcode.com/problems/permutations/description/
 *
 * algorithms
 * Medium (62.46%)
 * Likes:    3704
 * Dislikes: 104
 * Total Accepted:    590.5K
 * Total Submissions: 945.1K
 * Testcase Example:  '[1,2,3]'
 *
 * Given a collection of distinct integers, return all possible permutations.
 * 
 * Example:
 * 
 * 
 * Input: [1,2,3]
 * Output:
 * [
 * ⁠ [1,2,3],
 * ⁠ [1,3,2],
 * ⁠ [2,1,3],
 * ⁠ [2,3,1],
 * ⁠ [3,1,2],
 * ⁠ [3,2,1]
 * ]
 * 
 * 
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
		vector<vector<int>> result;
		iterate(result, nums, 0);
		return result;
    }

	void iterate(vector<vector<int>>& result, vector<int>& nums, int idx) {
		if (idx == nums.size()) {
			result.push_back(nums);
			return;
		}

		for (int i=idx; i<nums.size(); i++) {
			swap(nums[idx], nums[i]);
			iterate(result, nums, idx + 1);
			swap(nums[idx], nums[i]);
		}
	}
};
// @lc code=end
